frontend/src/components/wallet/WalletConnectionStatus.tsx

```diff
@@ -0,0 +1,142 @@
+
+declare const window: Window &
+  typeof globalThis & {
+    ethereum: any;
+  };
+
+interface MetaMaskError {
+  code: number;
+  message: string;
+}
+
+export const WalletConnectionStatus = () => {
+  const [errorMessage, setErrorMessage] = useState('');
+  const [defaultAccount, setDefaultAccount] = useState('');
+  const [shortAddress, setShortAddress] = useState('');
+  const { walletConnection, setWalletConnection } = useContext(WalletContext);
+
+  useEffect(() => {
+    // on load, try and connect the current wallet
+    initWallet();
+  });
+
+  const initWallet = () => {
+    if (window.ethereum) {
+      window.ethereum
+        .request({ method: 'eth_accounts' })
+        .then((result: Array<string>) => {
+          if (result.length > 0) {
+            accountChangedHandler(result[0]);
+          }
+        })
+        .catch((error: MetaMaskError) => {
+          // Some unexpected error.
+          console.error(`Metamask returned error: ${error.message}`);
+          setErrorMessage(error.message);
+        });
+
+      // handle account change
+      window.ethereum.on('accountsChanged', handleAccountsChanged);
+    }
+  };
+
+  function handleAccountsChanged(accounts: Array<string>) {
+    if (accounts.length === 0) {
+      // MetaMask is locked or the user has not connected any accounts
+      setErrorMessage('MetaMask has disconnected');
+      accountChangedHandler('');
+    } else if (accounts[0] !== defaultAccount) {
+      accountChangedHandler(accounts[0]);
+    }
+  }
+
+  const connectWalletHandler = () => {
+    if (window.ethereum) {
+      setErrorMessage('');
+      window.ethereum
+        .request({ method: 'eth_requestAccounts' })
+        .then((result: Array<string>) => {
+          if (result.length > 0) {
+            accountChangedHandler(result[0]);
+          }
+        })
+        .catch((error: MetaMaskError) => {
+          let msg = error.message;
+          if (error.code === 4001) {
+            msg = 'You canceled the request. Please try again.';
+          }
+          setErrorMessage(msg);
+        });
+    } else {
+      // MetaMask is not installed
+      setErrorMessage('Please install MetaMask');
+    }
+  };
+
+  const accountChangedHandler = (newAccount: string) => {
+    if (newAccount !== defaultAccount) {
+      setDefaultAccount(newAccount);
+
+      if (newAccount.length > 0) {
+        setShortAddress(shorten(utils.getAddress(newAccount)));
+        setWalletConnection({
+          connected: true,
+          account: newAccount,
+          provider: 'metamask',
+        });
+      } else {
+        setShortAddress('');
+        setWalletConnection({
+          connected: false,
+          account: '',
+          provider: '',
+        });
+      }
+    }
+  };
+
+  const shorten = (fullAddress: string) => {
+    return (
+      fullAddress.substring(0, 6) +
+      '...' +
+      fullAddress.substring(fullAddress.length - 4)
+    );
+  };
+
+  return (
+    <div>
```

Comment:
Need the extra div? Might be able to just do a fragment `<> </>`

Reply:
Nope.  Don't need it.  I fragmentized it.